packages/@react-aria/virtualizer/src/ScrollView.tsx

```diff
@@ -177,27 +182,38 @@ export function useScrollView(props: ScrollViewProps, ref: RefObject<HTMLElement
         });
       }
     }
+
+    isUpdatingSize.current = false;
   });
 
-  let didUpdateSize = useRef(false);
+  // Update visible rect when the content size changes, in case scrollbars need to appear or disappear.
+  let lastContentSize = useRef<Size | null>(null);
   useLayoutEffect(() => {
-    // React doesn't allow flushSync inside effects, so queue a microtask.
-    // We also need to wait until all refs are set (e.g. when passing a ref down from a parent).
-    queueMicrotask(() => {
-      if (!didUpdateSize.current) {
-        didUpdateSize.current = true;
-        updateSize(flushSync);
+    if (!isUpdatingSize.current) {
+      // Detect when scrollbar state might have changed.
+      let prevShowX = lastContentSize.current?.width > state.width;
+      let prevShowY = lastContentSize.current?.height > state.height;
+      let curShowX = contentSize.width > state.width;
+      let curShowY = contentSize.height > state.height;
+
+      if (lastContentSize.current && prevShowX === curShowX && prevShowY === curShowY) {
+        return;
+      }
+
+      // React doesn't allow flushSync inside effects, so queue a microtask.
+      // We also need to wait until all refs are set (e.g. when passing a ref down from a parent).
+      // If we are in an `act` environment, update immediately without a microtask so you don't need
+      // to mock timers in tests. In this case, the update is synchronous already.
+      if (globalThis.IS_REACT_ACT_ENVIRONMENT) {
```

Comment:
possibly not going to work, as some Issues out there recommend setting this to false in the tests... 
https://github.com/testing-library/react-testing-library/issues/1108
I can't imagine this is common, but I imagine this will be fun for people to debug, hahaimgonnagocrynow